### PR TITLE
SI-8578 Avoid fresh name clashes under -Ydelambdafy:method

### PR DESCRIPTION
**Pull request by retronym**
Review by @lrytz

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/Delambdafy.scala [hunk 1]
```diff
@@ -238,7 +238,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
         //      - reinstate the assertion in `Erasure.resolveAnonymousBridgeClash`
         val suffix = "$lambda$" + (
           if (funOwner.isPrimaryConstructor) ""
-          else "$" + funOwner.name
+          else "$" + funOwner.name + "$"
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Delambdafy.scala:241]: @lrytz Checkout the lines just above this for some ideas I left last time I reviewed this code. You might want to revisit this area in your backend work.

File: test/files/pos/t8578.scala [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class DuplicateClassName {
+  () => {
+    {() => ()}
```
Comment (adriaanm) [inline test/files/pos/t8578.scala:3]: I feel like it's looking straight back at me, no matter how I move my head!

Comment (som-snytt) [inline test/files/pos/t8578.scala:3]: :smile:

**Comment by lrytz** [comment]
LGTM

File: src/compiler/scala/tools/nsc/transform/Delambdafy.scala (hunk 1 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Delambdafy.scala:241]: I'll add it to my list, thanks.

**Comment by retronym** [comment]
I'm running a local community build with `#2.11.x` and https://github.com/typesafehub/community-builds/pull/39 and it spotted this bug, too:

```
[info] [error] Error while emitting QueryableTest.scala
[info] [error] assertion failed: Maps `symExemplars` and `exemplars` got out of synch.
[info] [warn] /Users/jason/code/community-builds/target-0.8.1/project-builds/slick-03f0024b40c1c01b119db099aa80b46eb6a8d5a5/slick-testkit/src/test/scala/scala/slick/test/direct/QueryableTest.scala:291: Class scala.slick.test.direct.QueryableTest$lambda$$$anonfun$612 differs only in case from scala.slick.test.direct.QueryableTest$lambda$$$anonfun$612. Such classes will overwrite one another on case-insensitive filesystems.
[info] [warn]         val inMemResult = for( o <- inMem; i <- inMem; if i.sales == o.sales ) yield i.name
[info] [warn]                                              ^
[info] [warn] /Users/jason/code/community-builds/target-0.8.1/project-builds/slick-03f0024b40c1c01b119db099aa80b46eb6a8d5a5/slick-testkit/src/test/scala/scala/slick/test/direct/QueryableTest.scala:291: Class scala.slick.test.direct.QueryableTest$lambda$$$anonfun$611 differs only in case from scala.slick.test.direct.QueryableTest$lambda$$$anonfun$611. Such classes will overwrite one another on case-insensitive filesystems.
[info] [warn]         val inMemResult = for( o <- inMem; i <- inMem; if i.sales == o.sales ) yield i.name
[info] [warn]                                                                   ^
[info] [warn] four warnings found
[info] [error] three errors found
[info] [error] (testkit/test:compile) Compilation failed
```

**Comment by lrytz** [comment]
i don't see any difference, not even in case..

**Comment by retronym** [comment]
Yeah, the warning is assuming that the only way to run into clashing class names is with user-level problems like `class C; class c`, but it still fires for in this case when the compiler generates the name clash.

**Comment by lrytz** [comment]
Makes sense.

**Pull request merged**